pkg/store/bucket.go

```diff
@@ -1320,23 +1324,72 @@ func (r *bucketIndexReader) ExpandedPostings(ms []*labels.Matcher) ([]uint64, er
 			return nil, errors.Wrap(err, "toPostingGroup")
 		}
 
+		// Intersection with empty postings would return no postings anyway.
+		if pg.alwaysEmptyPostings() {
+			return nil, nil
+		}
+
 		postingGroups = append(postingGroups, pg)
+		allRequested = allRequested || pg.addAll
+		hasAdds = hasAdds || len(pg.addKeys) > 0
+
+		// Postings returned by fetchPostings will be in the same order as keys
+		// so it's important that we iterate them in the same order later,
+		// since we don't build any label -> keys index map.
+		keys = append(keys, pg.addKeys...)
+		keys = append(keys, pg.removeKeys...)
 	}
 
 	if len(postingGroups) == 0 {
 		return nil, nil
 	}
 
-	if err := r.fetchPostings(postingGroups); err != nil {
+	allKeyIndex := -1
+	// we only need All postings if there are no other adds. If there are, we can skip fetching ALL postings completely.
+	if allRequested && !hasAdds {
+		// Remember the index (will be used later as a flag, and also to access postings),
+		// and ask fetchPostings to fetch ALL postings too.
+		allKeyIndex = len(keys)
+		keys = append(keys, getAllPostingsKeyLabel())
+	}
+
+	fetchedPostings, err := r.fetchPostings(keys)
+	if err != nil {
 		return nil, errors.Wrap(err, "get postings")
 	}
 
```

Comment:
We may add a check here to ensure that `len(fetchedPostings) == len(keys)`. If not, return an error.

Reply:
That is the contract of the function and is easy to verify, since the first thing it does is to allocate the output slice of the same length. If someone breaks that contract with his changes, I think it's OK if this code starts crashing.

Reply:
I will at least add a comment about it to `fetchPostings` though.